### ISSUE TITLE
How do I perform nested switching

### ISSUE THREAD

**Issue by kshenoy** [AndrewRadev/switch.vim#31]
I wish to switch between 3 strings: '*', '* Y' and '* N'

I tried the following settings:

    let g:switch_find_smallest_match = 0
    let g:switch_custom_definitions = [['* Y', '* N', '*']]

but this doesn't seem to work. What am I missing here?

**Comment by AndrewRadev** [comment]
Sorry it took a while to get to this.

The problem here is that this kind of format, with a simple list, doesn't work that well when the words have overlap -- the shortest match takes effect :/. One way you can get it to work is like this:

``` vim
let g:switch_find_smallest_match = 0
let g:switch_custom_definitions = [
      \   {'\* Y': '\* N'},
      \   {'\* N': '\*'},
      \   {'\*':   '\* Y'}
      \ ]
```

Each entry is a single attempt at a match, and they're tried in order due to `find_smallest_match = 0`. I know it's not ideal, but it's difficult to make the logic flexible enough to handle most use cases. I've got some plans for that, but they need some rearchitecturing.

For now, does this take care of your particular use case?

**Comment by kshenoy** [comment]
> the shortest match takes effect 

I thought that was why we have to use `let g:switch_find_smallest_match = 0` so that shortest match doesn't take effect. I'm a bit confused about the role of that variable now.

> For now, does this take care of your particular use case?

Yes, perfectly. Thanks for the great plugin :)

**Comment by AndrewRadev** [comment]
> Yes, perfectly. Thanks for the great plugin :)

Thank you for the kind words :).

> I thought that was why we have to use let g:switch_find_smallest_match = 0 so
> that shortest match doesn't take effect. I'm a bit confused about the role of
> that variable now.

That variable controls whether the full list of switches is attempted in-order, or whether it should try to find the "best" match. Unfortunately, the list `['*', '*Y', '*N']` is transformed into something like:

``` vim
\   {
\     '\* Y': '\* N',
\     '\* N': '\*',
\     '\*':   '\* Y',
\   }
```

And that dictionary is unsorted, which means it's impossible to apply an ordering to it. As you'll notice, the workaround I provided has a list of three hashes. And that works :). And I've just pushed a commit that should transform the short-hand list style into separate hashes for each pattern.

Long story short, if you pull now, your original attempt should work:

``` vim
let g:switch_find_smallest_match = 0
let g:switch_custom_definitions = [['* Y', '* N', '*']]
```

Try it out and let me know.

**Comment by AndrewRadev** [comment]
Since you issue seems to be fixed either way, I guess there's no need to keep this issue open. Feel free to reopen it if you run into issues or something.
